# Wrapping my head around the Gopher Pipeline RNAseq scripts

This'll hold my attempts to understand what's going on with the RNAseq portion of the [Gopher Pipelines](https://bitbucket.org/jgarbe/gopher-pipelines).

### Location of the scripts on MSI
```
/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/bin/
```

### Space-station view of what's happening.

The ```rnaseq-pipeline``` bash script runs the ```rnaseq-pipeline.pl``` script, which does lotsa stuff, as well as calling the ```rnaseq-singlesample.pl``` script.


### Relationship to the ```Pipelinev4.pm``` module file
The ```rnaseq-pipeline.pl``` uses the following subroutines, mostly as report generation helpers, and none of which do any analyses (as far as I can tell). Found like so

```
grep 'sub ' Pipelinev4.pm | cut -d' ' -f2 | grep -v '^$' | awk '{print$0"("}' | grep -o -f - rnaseq-pipeline.pl | sort | uniq | sed 's/(//g' > module_subs_in_rnaseq-pipeline.txt
```

The rnaseq-singlesample.pl is where the exciting stuff happens. We find the subroutines as above

```
grep 'sub ' Pipelinev4.pm | cut -d' ' -f2 | grep -v '^$' | awk '{print$0"("}' | grep -o -f - rnaseq-singlesample.pl | sort | uniq | sed 's/(//g' > module_subs_in_rnaseq-singlesample.txt

cat module_subs_in_rnaseq-* | sort | uniq > module_subs4rnaseq.txt
```

The set of subroutines defined in ```Pipelinev4.pm``` and present in ```rnaseq-pipeline.pl``` or ```rnaseq-singlesample.pl``` is listed below.


Sub | routine | names |
-----|-----|-----
`allmetrics` | `bowtie2indexcheck` | `cleansam`
`compilefolder` | `cuffquant` | `diegracefully`
`expressiontableplot` | `fastqc` | `fastqcplot`
`featurecounts` | `fieldlist` | `getsinglesamplestats`
`h` | `h1` | `hisat2`
`hisat2indexcheck` | `hisat2plot` | `image`
`insertsize` | `insertsizeplot` | `line`
`programversions` | `qc` | `readinextraoptions`
`requiredprograms` | `run` | `samplesheetandfoldersetup`
`samtoolssortandindexbam` | `scratchfoldersetup` | `sphinx`
`startreport` | `subsample` | `subsampleplot`
`tophat2` | `tophat2plot` | `trimmomatic`
`writestats` |  |


### Other ```perl``` scripts run within the script

```
rnaseq-singlesample.pl #/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/bin/

cufftablerename.pl #/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/software/gopher-biotools/

subread-merge.pl #/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/software/gopher-biotools/
```



### Workflow of function calls in ```rnaseq-pipeline.pl```

- Initialize with options [```&init(\%samples, \%args, \%stats);```]
  - Where these options are read from I can't tell?

- Begin the report [```startreport(\%samples, \%args, \%stats, "RNA-Seq Analysis Report");```]

- Run the per sample commands [```&singlesamples(\%samples, \%args, \%stats);```]
  - This is where options are processed and turned into arguments to ```rnaseq-singlesample.pl```
  - The command is written to a file, which is then ```cat```-ed into ```parallel```
  - The stats from the runs are then collected.

- Normalize counts with ```cuffnorm``` [```&cuffnorm(\%samples, \%args, \%stats);```]
  - The ```.cxb``` files this function is looking for are actually generated by ```cuffquant```, which is called in ```rnaseq-singlesample.pl```
  - Includes a of renaming of `.fkpm_table` file 
    - `/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/software/gopher-biotools/cufftablerename.pl`

- The differential expression function is commented out in the copy I have... [```#&differentialexpression(\%samples, \%args, \%stats);```]

- Run the all sample commands [```&allsample(\%samples, \%args, \%stats);```]
  - Generate plots
    - ```subsampleplot($samples, $args, $stats);```
    - ```fastqcplot($samples, $args, $stats);```
    - ```hisat2plot($samples, $args, $stats);``` or ```tophat2plot($samples, $args, $stats);```
    - ```insertsizeplot($samples, $args, $stats);```
  - Make directories and move stuff into them
    - lotsa `compilefolder()` calls as well as some `cp` and `tar`-ing
  - Merge subreads
    - `/panfs/roc/groups/3/umii/public/gopher-pipelines/dev/software/gopher-biotools/subread-merge.pl`
  - Per-sample count files for Sue Rathe
    - Seems to just involve some ```cut``` operations to split apart a file into samples.

- Finish report [```&finishreport(\%samples, \%args, \%stats);```]
  - Heavy use of helper functions from ```PipelineV4.pm``` for report generation

- Collect metrics [```allmetrics(\%samples, \%args, \%stats);```]

- Write the report [```sphinx(\%samples, \%args, \%stats, "RNA-Seq Analysis", "index.html");```] 

### Workflow of function calls in ```rnaseq-singlesample.pl```
- Initialize with options [```&init(\%args, \%stats);```]

- If the quality control flag is present
  - Run ```trimmomatic``` [``` trimmomatic(\%args, \%stats);```]
  - followed by ```fastqc``` [```fastqc(\%args, \%stats, "fastqc-trim");```]

- Conditional upon which index is present
  - Run ```Hisat2``` [```hisat2(\%args, \%stats);```]
  - or ```Tophat2``` [```tophat2(\%args, \%stats);```]

- ```Picard cleansam``` is next but is commented out. [```#&cleansam(\%args, \%stats);```]

- ```.bam``` sorting and indexing [```samtoolssortandindexbam(\%args, \%stats);```]

- Collect insert size metrics [```insertsize(\%args, \%stats);```]

- Feature counts on subreads [```featurecounts(\%args, \%stats);```]

- ```Cuffquant``` [```cuffquant(\%args, \%stats);```]

- Stats writing [```writestats(\%args, \%stats);```]

- Finally, `touch $args{scratchfolder}/Complete`;
